File: VisualNovelEngine/cpp/src/core/Engine.cpp
#include "core/Engine.h"
#include "core/EventManager.h"
#include "scene/SceneManager.h"
#include "rendering/RenderEngine.h"
#include "resources/ResourceManager.h"
#include "script/ScriptEngine.h"
#include <SFML/System/Clock.hpp>
#include <iostream>

namespace VNE {

Engine* Engine::s_instance = nullptr;

Engine::Engine() {
    s_instance = this;
}

Engine::~Engine() {
    shutdown();
    s_instance = nullptr;
}

bool Engine::initialize(const GameConfig& config) {
    if (m_initialized) {
        return true;
    }
    
    m_config = config;
    
    // Initialize subsystems
    m_resourceManager = std::make_unique<ResourceManager>();
    if (!m_resourceManager->initialize()) {
        std::cerr << "Failed to initialize ResourceManager" << std::endl;
        return false;
    }
    
    m_renderEngine = std::make_unique<RenderEngine>();
    if (!m_renderEngine->initialize(config.width, config.height, config.title, config.vsync)) {
        std::cerr << "Failed to initialize RenderEngine" << std::endl;
        return false;
    }
    
    m_eventManager = std::make_unique<EventManager>();
    if (!m_eventManager->initialize()) {
        std::cerr << "Failed to initialize EventManager" << std::endl;
        return false;
    }
    
    m_sceneManager = std::make_unique<SceneManager>();
    if (!m_sceneManager->initialize()) {
        std::cerr << "Failed to initialize SceneManager" << std::endl;
        return false;
    }
    
    m_scriptEngine = std::make_unique<ScriptEngine>();
    if (!m_scriptEngine->initialize()) {
        std::cerr << "Failed to initialize ScriptEngine" << std::endl;
        return false;
    }
    
    m_initialized = true;
    std::cout << "Engine initialized successfully" << std::endl;
    return true;
}

void Engine::shutdown() {
    if (!m_initialized) {
        return;
    }
    
    m_running = false;
    
    if (m_scriptEngine) {
        m_scriptEngine->shutdown();
        m_scriptEngine.reset();
    }
    
    if (m_sceneManager) {
        m_sceneManager->shutdown();
        m_sceneManager.reset();
    }
    
    if (m_eventManager) {
        m_eventManager->shutdown();
        m_eventManager.reset();
    }
    
    if (m_renderEngine) {
        m_renderEngine->shutdown();
        m_renderEngine.reset();
    }
    
    if (m_resourceManager) {
        m_resourceManager->shutdown();
        m_resourceManager.reset();
    }
    
    m_initialized = false;
    std::cout << "Engine shutdown complete" << std::endl;
}

void Engine::run() {
    if (!m_initialized) {
        std::cerr << "Engine not initialized" << std::endl;
        return;
    }
    
    m_running = true;
    sf::Clock deltaClock;
    
    while (m_running && m_renderEngine->isOpen()) {
        float deltaTime = deltaClock.restart().asSeconds();
        
        processEvents();
        update(deltaTime);
        render();
    }
}

void Engine::stop() {
    m_running = false;
}

void Engine::processEvents() {
    m_eventManager->processEvents(m_renderEngine->getWindow());
    
    // Check for window close
    if (!m_renderEngine->isOpen()) {
        m_running = false;
    }
}

void Engine::update(float deltaTime) {
    m_sceneManager->update(deltaTime);
}

void Engine::render() {
    m_renderEngine->clear();
    m_sceneManager->render();
    m_renderEngine->display();
}

} // namespace VNE

File: VisualNovelEngine/rust/src/core/engine.rs
use crate::core::config::GameConfig;
use crate::rendering::RenderEngine;
use crate::resources::ResourceManager;
use crate::scene::SceneManager;
use crate::script::ScriptEngine;
use anyhow::Result;
use std::sync::Arc;
use parking_lot::Mutex;

/// Main engine struct that coordinates all subsystems
pub struct Engine {
    config: GameConfig,
    resource_manager: Arc<Mutex<ResourceManager>>,
    render_engine: Arc<Mutex<RenderEngine>>,
    scene_manager: Arc<Mutex<SceneManager>>,
    script_engine: Arc<Mutex<ScriptEngine>>,
    running: bool,
}

impl Engine {
    /// Create a new engine instance
    pub fn new() -> Self {
        Self {
            config: GameConfig::default(),
            resource_manager: Arc::new(Mutex::new(ResourceManager::new())),
            render_engine: Arc::new(Mutex::new(RenderEngine::new())),
            scene_manager: Arc::new(Mutex::new(SceneManager::new())),
            script_engine: Arc::new(Mutex::new(ScriptEngine::new())),
            running: false,
        }
    }

    /// Initialize the engine with the given configuration
    pub fn initialize(&mut self, config: GameConfig) -> Result<()> {
        log::info!("Initializing Visual Novel Engine...");
        
        self.config = config;

        // Initialize resource manager
        {
            let mut rm = self.resource_manager.lock();
            rm.initialize()?;
        }

        // Initialize render engine
        {
            let mut re = self.render_engine.lock();
            re.initialize(
                self.config.width,
                self.config.height,
                &self.config.title,
                self.config.vsync,
            )?;
        }

        // Initialize scene manager
        {
            let mut sm = self.scene_manager.lock();
            sm.initialize()?;
        }

        // Initialize script engine
        {
            let mut se = self.script_engine.lock();
            se.initialize()?;
        }

        log::info!("Engine initialized successfully");
        Ok(())
    }

    /// Shutdown the engine
    pub fn shutdown(&mut self) {
        log::info!("Shutting down engine...");
        self.running = false;
        
        // Shutdown subsystems in reverse order
        self.script_engine.lock().shutdown();
        self.scene_manager.lock().shutdown();
        self.render_engine.lock().shutdown();
        self.resource_manager.lock().shutdown();
        
        log::info!("Engine shutdown complete");
    }

    /// Run the main game loop
    pub fn run(&mut self) -> Result<()> {
        if self.running {
            return Ok(());
        }

        self.running = true;
        log::info!("Starting main game loop");

        // Main loop would go here - simplified for this example
        while self.running {
            self.update(1.0 / 60.0)?;
            self.render()?;
        }

        Ok(())
    }

    /// Stop the engine
    pub fn stop(&mut self) {
        self.running = false;
    }

    /// Update all subsystems
    fn update(&mut self, delta_time: f32) -> Result<()> {
        self.scene_manager.lock().update(delta_time);
        Ok(())
    }

    /// Render the current frame
    fn render(&mut self) -> Result<()> {
        self.render_engine.lock().clear();
        self.scene_manager.lock().render();
        self.render_engine.lock().present()?;
        Ok(())
    }

    // Getters for subsystems
    pub fn resource_manager(&self) -> Arc<Mutex<ResourceManager>> {
        Arc::clone(&self.resource_manager)
    }

    pub fn render_engine(&self) -> Arc<Mutex<RenderEngine>> {
        Arc::clone(&self.render_engine)
    }

    pub fn scene_manager(&self) -> Arc<Mutex<SceneManager>> {
        Arc::clone(&self.scene_manager)
    }

    pub fn script_engine(&self) -> Arc<Mutex<ScriptEngine>> {
        Arc::clone(&self.script_engine)
    }
}

impl Default for Engine {
    fn default() -> Self {
        Self::new()
    }
}

impl Drop for Engine {
    fn drop(&mut self) {
        if self.running {
            self.shutdown();
        }
    }
}
